angr/analyses/propagator/propagator.py

```diff
@@ -0,0 +1,281 @@
+                    state.local_variables[offset] = value
+                else:
+                    if state.local_variables[offset] != value:
+                        state.local_variables[offset] = TOP
+
+        return state
+
+    def store_local_variable(self, offset, size, value):  # pylint:disable=unused-argument
+        # TODO: Handle size
+        self.local_variables[offset] = value
+
+    def load_local_variable(self, offset, size):  # pylint:disable=unused-argument
+        # TODO: Handle size
+        try:
+            return self.local_variables[offset]
+        except KeyError:
+            return BOTTOM
+
+    def store_register(self, offset, size, value):
+        if size != self.gpr_size:
+            return
+
+        self.registers[offset] = value
+
+    def load_register(self, offset, size):
+
+        # TODO: Fix me
+        if size != self.gpr_size:
+            return BOTTOM
+
+        try:
+            return self.registers[offset]
+        except KeyError:
+            return BOTTOM
+
+# AIL state
+
+class PropagatorAILState(PropagatorState):
+    def __init__(self, arch, replacements=None):
+        super().__init__(arch, replacements=replacements)
+
+        self._variables = { }  # variable to values
+
+    def __repr__(self):
+        return "<PropagatorAILState>"
+
+    def copy(self):
+        rd = PropagatorAILState(
+            self.arch,
+            replacements=self._replacements.copy(),
+        )
+
+        rd._variables = self._variables.copy()
+
+        return rd
+
+    def merge(self, *others):
+        state = super().merge(*others)
+
+        for o in others:
+            for k, v in o._variables.items():
+                if k not in state._variables:
+                    state._variables[k] = v
+                else:
+                    if state._variables[k] != o._variables[k]:
+                        # Go to TOP
+                        state._variables[k] = TOP
+
+        return state
+
+    def store_variable(self, old, new):
+        if new is not None:
+            self._variables[old] = new
+
+    def get_variable(self, old):
+        return self._variables.get(old, None)
+
+    def remove_variable(self, old):
+        self._variables.pop(old, None)
+
+    def filter_variables(self, atom):
+        keys_to_remove = set()
+
+        for k, v in self._variables.items():
+            if isinstance(v, ailment.Expr.Expression) and (v == atom or v.has_atom(atom)):
+                keys_to_remove.add(k)
+
+        for k in keys_to_remove:
+            self._variables.pop(k)
+
+
+class PropagatorAnalysis(ForwardAnalysis, Analysis):  # pylint:disable=abstract-method
+    """
+    PropagatorAnalysis propagates values, either constants or variables, across a block or a function. It supports both
+    VEX and AIL. It performs certain arithmetic operations between constants, including but are not limited to:
+
+    - addition
+    - subtraction
+    - multiplication
+    - division
+    - xor
```

Comment:
one day I will understand why you need to implement every single operation for every single analysis instead of falling back to our current understanding of how to perform each of the vex/ail ops, but today is not that day